### PR TITLE
Fixed #28152 -- Changed SetSerializer to serialize to set literals.

### PR DESCRIPTION
**Pull request by jdufresne**
Creating this PR to resolve the discussion in #8432. The discussion revealed that the migration serializer renders sets as `set([...])`. When searching for code that could be converted to use set literals or set comprehension, these auto-generated code fragments surfaced.

<https://code.djangoproject.com/ticket/28152>

### CODE REVIEW

File: tests/migrations/test_serializer.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+        evaled = eval(value)
+        self.assertEqual(evaled, orig)
+
+    def test_serilizes_to_set_literal(self):
+        orig = {'a'}
+        serializer = SetSerializer(orig)
+        value, imports = serializer.serialize()
+        self.assertEqual(value, "{'a'}")
+        evaled = eval(value)
+        self.assertEqual(evaled, orig)
```
Comment (charettes) [inline tests/migrations/test_serializer.py]: Moving migration serializer tests to it's own module could make sense in a separate commit but let's keep these in `test_writer.WriterTests()` for now.

**Comment by jdufresne** [comment]
Thanks for the review. I've moved the tests to `test_writer.py` and simplified them by using the existing asserting functions. I've also created the Trac ticket.

File: django/db/migrations/serializer.py [hunk 2]
```diff
@@ -244,8 +244,9 @@ def _format(self):
 
 class SetSerializer(BaseSequenceSerializer):
     def _format(self):
-        # Don't use the literal "{%s}" as it doesn't support empty set
-        return "set([%s])"
+        # When value is empty, the empty set should be serialized as "set()",
+        # not "{}" because {} is an empty dict.
+        return '{%s}' if self.value else 'set(%s)'
```
Comment (jarshwah) [inline django/db/migrations/serializer.py]: I think this actually makes the code look worse in this particular case. Is there any tangible benefit?

Comment (jdufresne) [inline django/db/migrations/serializer.py]: In PR #8432, the goal is to replace the pattern `set(<generator expression>)` with set comprehension. As this pattern occasionally creeps in over time, regular maintenance may be required keep the preferred comprehension syntax. To solve this "over time" problem, @charettes suggested the possibility of using the static analysis tool [flake8-comprehensions](https://pypi.python.org/pypi/flake8-comprehensions/1.2.1) to find and report these patterns for us. One snag with using the tool, it flagged migration files as having suboptimal set syntax `set([...])` instead of set literals.

So the goal of this change is to make the `SetSerializer` friendlier for static analysis.

Comment (charettes) [inline django/db/migrations/serializer.py]: The other alternative would be to avoid performing analysis of the migration files.

FWIW I personally find set literals to be better looking and easier to parse particularly in cases where the set doesn't contain many element and the `set([])` boilerplate accounts for half the length of the whole statement.

Comment (MarkusH) [inline django/db/migrations/serializer.py]: We should remember that migration files are auto generated code. They first and foremost need to not have syntax errors.

There was a discussion on running some form of isort on migration files while generating them. We decided against that for the sake of migration files being auto-generated. Furthermore, lines in migration files can easily end up having lines way past 79 or 119 characters. And I'm against starting to linebreak those lines with some other tool. Users should exclude migration files from linting, IMO.

Comment (jdufresne) [inline django/db/migrations/serializer.py]: In the projects I work on, migrations often contain more than auto-generated code. True, many start off as auto-generated code, but between schema tweaks and data migrations, there is plenty of written code. For this reason, my projects have success including `migrations` directories in flake8 runs.

Comment (charettes) [inline django/db/migrations/serializer.py]: > We should remember that migration files are auto generated code. They first and foremost need to not have syntax errors.

Do you consider the usage of set literals more likely to cause syntax errors?

Comment (MarkusH) [inline django/db/migrations/serializer.py]: If we were to change this --- I don't have strong opinions on that --- I don't want to see this as a precedent for making migration files to adhere to flake8.

Given this particular line, it makes the code _slightly_ harder to read and needs a second line of comment. _I_ don't see the benefit here.

My 2 ¢

Comment (jdufresne) [inline django/db/migrations/serializer.py]: >  I don't want to see this as a precedent for making migration files to adhere to flake8.

They are already included in runs of flake8. [See `setup.cfg`  where there is no exclude rule for `migrations`](https://github.com/django/django/blob/master/setup.cfg#L6).

This is the reason this change was originally discussed. If flake8 were to check for set literals, something would need to change with regards to migrations. Either:

* Change migrations to use set literals (this PR)
* Ignore `migrations` in flake8 (sounds like you're suggesting this alternative)

I still think there is some merit to running flake8 on migration files as they often contain written code.

Due to other reasons, the addition of flake8-comprensions won't happen for the time being, so there is no longer a need for this change. We could consider simply closing it if there is not agreement.

Comment (timgraham) [inline django/db/migrations/serializer.py]: I think the change is nice since set literals are less verbose (less brackets/parenthesis to visually parse). Josh, do you still have a concern?

Comment (jarshwah) [inline django/db/migrations/serializer.py]: In general I agree. It was only this specific case I had issues with because it introduces a branch which is noisier than the syntax we're trying to standardize on. That said it's in a codegen path, so the nicer output is likely to outweigh messier input. 

No objections to merge.
